homeassistant/components/number/__init__.py

```diff
@@ -419,3 +420,47 @@ def _report_deprecated_number_entity(self) -> None:
                 type(self),
                 report_issue,
             )
+
+
+@dataclasses.dataclass
+class NumberExtraStoredData(ExtraStoredData):
+    """Object to hold extra stored data."""
+
+    native_max_value: float | None
+    native_min_value: float | None
+    native_step: float | None
+    native_unit_of_measurement: str | None
+    native_value: float | None
+
+    def as_dict(self) -> dict[str, Any]:
+        """Return a dict representation of the number data."""
+        return dataclasses.asdict(self)
+
+    @classmethod
+    def from_dict(cls, restored: dict[str, Any]) -> NumberExtraStoredData | None:
+        """Initialize a stored number state from a dict."""
+        try:
+            return cls(**restored)
```

Comment:
Specify the exact keys. Otherwise this prevents a rollback from a version that has added keys.